### ISSUE TITLE
Accessing name of indexed component element is slow for later elements

### ISSUE THREAD

**Issue by mfripp** [Pyomo/pyomo#2560]
I have some troubleshooting code where I need to show the names of a bunch of constraint objects (i.e., constraint name plus index value). `constraint_element.name` is handy for this, but I have found that accessing this is very slow (up to 400 ms) when there are many elements in the indexed constraint, and the access is especially slow for later elements.

Here is some code that demonstrates this. This creates an indexed constraint with 20000 elements (you may need to raise or lower it depending on the speed of your computer). Then it accesses the name of each element, in groups of 1000 from start to end, then back from end to start. This would ideally be nearly instantaneous for all elements. But on my computer, the first 1000 constraint elements take around 0.2 ms each, while the last 1000 take around 6 ms each.

```
import pyomo.environ as pyo
import time

m = pyo.ConcreteModel()
m.S = pyo.Set(initialize=range(20000))
m.V = pyo.Var(m.S)
m.C = pyo.Constraint(m.S, rule=lambda m, s: m.V[s] >= s)

t = time.time()
last_shown = 0
cur = 0
# test access speed from start to end
for k, c in m.C.items():
    cur += 1
    x = c.name  # retrieve constraint name
    if cur == last_shown + 1000:
        dur = time.time() - t
        print(f"constraints {last_shown + 1} - {cur}: {dur:0.1f} ms each")
        t = time.time()
        last_shown = cur

# test access speed from end to start
print()
for k, c in reversed(list(m.C.items())):
    cur -= 1
    x = c.name  # retrieve constraint name
    if cur == last_shown - 1000:
        dur = time.time() - t
        print(f"constraints {cur + 1} - {last_shown}: {dur:0.1f} ms each")
        t = time.time()
        last_shown = cur
```

Output from this code (expected constant, low access times for all constraints, but it is quite high for high-numbered constraints):

```
constraints 1 - 1000: 0.2 ms each
constraints 1001 - 2000: 0.4 ms each
constraints 2001 - 3000: 0.8 ms each
constraints 3001 - 4000: 1.0 ms each
constraints 4001 - 5000: 1.4 ms each
constraints 5001 - 6000: 2.1 ms each
constraints 6001 - 7000: 2.4 ms each
constraints 7001 - 8000: 2.6 ms each
constraints 8001 - 9000: 3.0 ms each
constraints 9001 - 10000: 2.9 ms each
constraints 10001 - 11000: 3.2 ms each
constraints 11001 - 12000: 3.5 ms each
constraints 12001 - 13000: 3.9 ms each
constraints 13001 - 14000: 4.2 ms each
constraints 14001 - 15000: 4.5 ms each
constraints 15001 - 16000: 4.8 ms each
constraints 16001 - 17000: 5.6 ms each
constraints 17001 - 18000: 5.6 ms each
constraints 18001 - 19000: 5.8 ms each
constraints 19001 - 20000: 6.2 ms each

constraints 19001 - 20000: 6.1 ms each
constraints 18001 - 19000: 5.7 ms each
constraints 17001 - 18000: 5.4 ms each
constraints 16001 - 17000: 5.2 ms each
constraints 15001 - 16000: 4.7 ms each
constraints 14001 - 15000: 4.6 ms each
constraints 13001 - 14000: 4.2 ms each
constraints 12001 - 13000: 3.8 ms each
constraints 11001 - 12000: 3.5 ms each
constraints 10001 - 11000: 3.1 ms each
constraints 9001 - 10000: 2.8 ms each
constraints 8001 - 9000: 2.5 ms each
constraints 7001 - 8000: 2.3 ms each
constraints 6001 - 7000: 2.0 ms each
constraints 5001 - 6000: 1.9 ms each
constraints 4001 - 5000: 1.6 ms each
constraints 3001 - 4000: 1.1 ms each
constraints 2001 - 3000: 0.8 ms each
constraints 1001 - 2000: 0.5 ms each
constraints 1 - 1000: 0.2 ms each
```

If you change `m.C` to `m.V` at the top of each loop and run it again, you will see that the same effect occurs for indexed variables as well.

In both cases, I'm guessing that `obj.name` scans the parent component until it finds the specified object, giving a quadratic time dependency for later objects, which can get very slow for components with many elements.

**Comment by emma58** [comment]
@mfripp, what version of Pyomo are you using?

**Comment by jsiirola** [comment]
@mfripp, this was expected behavior in older versions of Pyomo: we had prioritized memory footprint over the speed at which a component data object could access its index.  While this saved a measurable amount of memory, naive access of a component data's index (which `.name` does) resulted in a linear scan of the container's data dictionary to identify the index value.  There were ways to generate names in bulk more efficiently - and the assumption was that clients that were generating a large number of names would use the more efficient interface.

However, we kept running into situations where algorithms did not (or could not) make use of the efficient bulk generation of names -- resulting in significant performance impacts.  

Pyomo 6.4.1 changed the internal component data structure so that individual component datas now maintain pointers to their index value.  This change means that all `.name` accesses should be constant time.
